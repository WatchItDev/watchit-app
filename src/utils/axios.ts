import axios, { AxiosRequestConfig } from 'axios';
// config
import { HOST_API } from 'src/config-global';

// ----------------------------------------------------------------------

const axiosInstance = axios.create({ baseURL: HOST_API });

axiosInstance.interceptors.response.use(
  (res) => res,
  (error) => Promise.reject((error.response && error.response.data) || 'Something went wrong')
);

export default axiosInstance;

// ----------------------------------------------------------------------

export const fetcher = async (args: string | [string, AxiosRequestConfig]) => {
  const [url, config] = Array.isArray(args) ? args : [args];

  const res = await axiosInstance.get(url, { ...config });

  return res.data;
};

// ----------------------------------------------------------------------

export const endpoints = {
  chat: '/api/chat',
  kanban: '/api/kanban',
  calendar: '/api/calendar',
  auth: {
    me: '/api/auth/me',
    login: '/api/auth/login',
    register: '/api/auth/register',
  },
  mail: {
    list: '/api/mail/list',
    details: '/api/mail/details',
    labels: '/api/mail/labels',
  },
  post: {
    list: '/api/post/list',
    details: '/api/post/details',
    latest: '/api/post/latest',
    search: '/api/post/search',
  },
  product: {
    list: '/api/product/list',
    details: '/api/product/details',
    search: '/api/product/search',
  },
};
