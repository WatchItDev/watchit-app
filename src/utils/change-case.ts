export function paramCase(str: string) {
  return str
    .toLowerCase()
    .replace(/\s+/g, '-')
    .replace(/[^a-z0-9-]/g, '');
}

export function snakeCase(str: string) {
  return str
    .toLowerCase()
    .replace(/\s+/g, '_')
    .replace(/[^a-z0-9_]/g, '');
}
