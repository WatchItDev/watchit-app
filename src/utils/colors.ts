/**
 * Adjusts the brightness of a given color by darkening it.
 * The function decreases the RGB components of the color to create a darker shade.
 *
 * @param {string} color - A string representing the hexadecimal color value (e.g., "#RRGGBB").
 * @param {number} [tone=30] - The amount to darken each RGB component. Default is 30.
 * @returns {string} A new hexadecimal color string that represents the darker shade of the input color.
 */
export const darken = (color: string, tone = 30) => {
  let r = parseInt(color.slice(1, 3), 16);
  let g = parseInt(color.slice(3, 5), 16);
  let b = parseInt(color.slice(5, 7), 16);

  r = Math.max(0, r - tone);
  g = Math.max(0, g - tone);
  b = Math.max(0, b - tone);

  return `#${r.toString(16).padStart(2, '0')}${g.toString(16).padStart(2, '0')}${b.toString(16).padStart(2, '0')}`;
};
