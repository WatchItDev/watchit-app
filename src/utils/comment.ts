export const timeAgo = (date) => {
  const now = new Date();
  const seconds = Math.floor((now - date) / 1000);

  let interval = Math.floor(seconds / 31536000);
  if (interval >= 1) {
    return interval + 'y ago';
  }
  interval = Math.floor(seconds / 2592000);
  if (interval >= 1) {
    return interval + 'mo ago';
  }
  interval = Math.floor(seconds / 86400);
  if (interval >= 1) {
    return interval + 'd ago';
  }
  interval = Math.floor(seconds / 3600);
  if (interval >= 1) {
    return interval + 'h ago';
  }
  interval = Math.floor(seconds / 60);
  if (interval >= 1) {
    return interval + 'min ago';
  }
  return Math.floor(seconds) + 's ago';
};
