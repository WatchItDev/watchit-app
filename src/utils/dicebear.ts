export const dicebear = (seed: string) => {
  return `https://api.dicebear.com/9.x/bottts-neutral/svg?seed=${seed}`
}
