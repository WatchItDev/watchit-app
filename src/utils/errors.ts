/**
 * Enum of our known error names.
 */
export enum ERRORS {
  NOT_LOGGED_IN = 'NOT_LOGGED_IN',
  BUNDLER_UNAVAILABLE = 'BUNDLER_UNAVAILABLE',
  UNKNOWN_ERROR = 'UNKNOWN_ERROR',
}

/**
 * Error messages dictionary mapping our ERROR_NAMES to user-facing strings.
 */
export const ERROR_MESSAGES: Record<ERRORS, string> = {
  [ERRORS.NOT_LOGGED_IN]: 'Please log in to continue.',
  [ERRORS.BUNDLER_UNAVAILABLE]: 'The bundler is currently unavailable.',
  [ERRORS.UNKNOWN_ERROR]: 'An unknown error has occurred.',
};

let globalEnqueueSnackbar: ((message: string, options?: object) => void) | null = null;

/**
 * Allows us to set the globalEnqueueSnackbar function from our App (or any top-level component).
 * Must be called at least once (e.g., in App.tsx) so that notifyError() works.
 */
export function setGlobalErrorNotifier(
  enqueueSnackbarFn: (message: string, options?: object) => void
) {
  globalEnqueueSnackbar = enqueueSnackbarFn;
}

/**
 * Global function to notify an error by ERROR_NAMES.
 * If no message is found in the dictionary, it uses 'UNKNOWN_ERROR'.
 */
export function notifyError(errorName: ERRORS, fallbackMessage?: string) {
  if (!globalEnqueueSnackbar) {
    console.error('No globalEnqueueSnackbar is set. Cannot notify error.');
    return;
  }
  const message =
    ERROR_MESSAGES[errorName] ||
    fallbackMessage ||
    ERROR_MESSAGES[ERRORS.UNKNOWN_ERROR];

  globalEnqueueSnackbar(message, { variant: 'error' });
}
