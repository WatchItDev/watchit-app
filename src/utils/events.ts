import { Dispatch } from 'redux'
import { appendNotification } from '@src/redux/notifications'

export namespace Events {
  export const Handlers = (payload: any, profileId: string, dispatch?: Dispatch) => {
    const evenType: keyof typeof EventsHandlersMap = payload?.new?.payload?.type

    if (EventsHandlersMap[evenType]) {
      EventsHandlersMap[evenType](payload, profileId, dispatch)
    }
  }

  /**
   * Handle notification event for internal popover.
   * @param payload Data from the event (set as needed)
   * @param profileId Current user profile id for LENS Protocol
   * @param dispatch Redux dispatch function
   * @constructor
   */
  export const Notification = (payload: any, profileId: string, dispatch?: Dispatch) => {
    const notification = payload.new

    if (notification?.receiver_id === profileId) {
      if (dispatch) {
        dispatch(appendNotification(notification))
      }
    }
  }
  // @ts-ignore
  export const Transaction = (payload: any) => {}
  // @ts-ignore
  export const Publication = (payload: any) => {}
  // @ts-ignore
  export const Subscription = (payload: any) => {}
  // @ts-ignore
  export const Follower = (payload: any) => {}
}

export const EventsHandlersMap = {
  NOTIFICATION: Events.Notification,
  TRANSACTION: Events.Transaction,
  PUBLICATION: Events.Publication,
  SUBSCRIPTION: Events.Subscription,
  FOLLOWER: Events.Follower,
}
