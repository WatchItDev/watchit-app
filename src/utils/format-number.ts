import numeral from 'numeral'

type InputValue = string | number | null;

export function fNumber(number: InputValue) {
  return numeral(number).format()
}

export function fCurrency(number: InputValue) {
  const format = number ? numeral(number).format('$0,0.00') : ''

  return result(format, '.00')
}

export function fPercent(number: InputValue) {
  const format = number ? numeral(Number(number) / 100).format('0.0%') : ''

  return result(format, '.0')
}

export function fShortenNumber(number: InputValue) {
  const format = number ? numeral(number).format('0.00a') : ''

  return result(format, '.00')
}

export function fData(number: InputValue) {
  const format = number ? numeral(number).format('0.0 b') : ''

  return result(format, '.0')
}

function result(format: string, key = '.00') {
  const isInteger = format.includes(key)

  return isInteger ? format.replace(key, '') : format
}

export function formatBalanceNumber(balance: number) {
  const balanceOptions = { minimumFractionDigits: 1, maximumFractionDigits: 3 }
  return new Intl.NumberFormat('en-US', balanceOptions).format(balance as any)
}

interface AmountConstraintsProps {
  value: number;
  MAX_AMOUNT: number;
  MAX_POOL: number;
  setAmount: (value: number) => void;
  setCanContinue: (canContinue: boolean) => void;
}

export const handleAmountConstraints = ({
                                          value,
                                          MAX_AMOUNT,
                                          MAX_POOL,
                                          setAmount,
                                          setCanContinue,
                                        }: AmountConstraintsProps) => {
  if (value > MAX_POOL) {
    value = MAX_POOL // Truncate to a thousand millions
  }
  if (value < 0) {
    value = 0 // Set amount to 0 if lower than 0
  }
  setAmount(value)
  setCanContinue(value <= MAX_AMOUNT)

  // If amount is greater than balance, allow input but setCanContinue to false
  if (value > MAX_AMOUNT || value <= 0) {
    setCanContinue(false)
  }
  console.log('value', value)
}
