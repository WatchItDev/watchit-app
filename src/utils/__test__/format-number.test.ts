import { fNumber, fCurrency, fPercent, fShortenNumber, fData, formatBalanceNumber, handleAmountConstraints } from '@src/utils/format-number';

describe('[UTILS]: fNumber', () => {
  it('formats a regular number correctly', () => {
    expect(fNumber(1234)).toBe('1,234');
  });

  it('handles string numeric input', () => {
    expect(fNumber('5678')).toBe('5,678');
  });

  it('handles decimal numbers', () => {
    expect(fNumber(1234.56)).toBe('1,235');
  });

  it('handles negative numbers', () => {
    expect(fNumber(-1234)).toBe('-1,234');
  });

  it('handles null value', () => {
    expect(fNumber(null)).toBe('0');
  });

  it('handles zero', () => {
    expect(fNumber(0)).toBe('0');
  });

  it('matches snapshot for different inputs', () => {
    const results = [
      fNumber(1000),
      fNumber('2500'),
      fNumber(-350),
      fNumber(0),
      fNumber(null)
    ];
    expect(results).toMatchSnapshot();
  });
});

describe('[UTILS]: fCurrency', () => {
  it('formats currency for whole numbers', () => {
    expect(fCurrency(1000)).toBe('$1,000');
  });

  it('formats currency with decimal places', () => {
    expect(fCurrency(1000.5)).toBe('$1,000.50');
  });

  it('handles string numeric input', () => {
    expect(fCurrency('1000')).toBe('$1,000');
  });

  it('handles exact decimal values that should not show zeros', () => {
    expect(fCurrency(1000.00)).toBe('$1,000');
  });

  it('handles negative amounts', () => {
    expect(fCurrency(-500.75)).toBe('-$500.75');
  });

  it('handles null value', () => {
    expect(fCurrency(null)).toBe('');
  });

  it('handles zero value', () => {
    expect(fCurrency(0)).toBe('');
  });
});

describe('[UTILS]: fPercent', () => {
  it('formats percentage for whole numbers', () => {
    expect(fPercent(50)).toBe('50%');
  });

  it('formats percentage for decimal numbers', () => {
    expect(fPercent(25.5)).toBe('25.5%');
  });

  it('handles string numeric input', () => {
    expect(fPercent('75')).toBe('75%');
  });

  it('handles exact decimal values that should not show zeros', () => {
    expect(fPercent(100)).toBe('100%');
  });

  it('handles negative percentages', () => {
    expect(fPercent(-20)).toBe('-20%');
  });

  it('handles null value', () => {
    expect(fPercent(null)).toBe('');
  });

  it('handles zero value', () => {
    expect(fPercent(0)).toBe('');
  });
});

describe('[UTILS]: fShortenNumber', () => {
  it('shortens thousands', () => {
    expect(fShortenNumber(5000)).toBe('5k');
  });

  it('shortens millions', () => {
    expect(fShortenNumber(3500000)).toBe('3.50m');
  });

  it('shortens billions', () => {
    expect(fShortenNumber(1200000000)).toBe('1.20b');
  });

  it('handles decimal numbers', () => {
    expect(fShortenNumber(1234.56)).toBe('1.23k');
  });

  it('handles string numeric input', () => {
    expect(fShortenNumber('7500000')).toBe('7.50m');
  });

  it('handles null value', () => {
    expect(fShortenNumber(null)).toBe('');
  });

  it('handles zero value', () => {
    expect(fShortenNumber(0)).toBe('');
  });
});

describe('[UTILS]: fData', () => {
  it('formats bytes', () => {
    expect(fData(900)).toBe('900 B');
  });

  it('formats kilobytes', () => {
    expect(fData(1500)).toBe('1.5 KB');
  });

  it('formats megabytes', () => {
    expect(fData(2500000)).toBe('2.5 MB');
  });

  it('formats gigabytes', () => {
    expect(fData(3500000000)).toBe('3.5 GB');
  });

  it('handles string numeric input', () => {
    expect(fData('1048576')).toBe('1 MB');
  });

  it('handles null value', () => {
    expect(fData(null)).toBe('');
  });

  it('handles zero value', () => {
    expect(fData(0)).toBe('');
  });
});

describe('[UTILS]: formatBalanceNumber', () => {
  it('formats whole number with minimum one decimal place', () => {
    expect(formatBalanceNumber(100)).toBe('100.0');
  });

  it('formats decimal number with up to three decimal places', () => {
    expect(formatBalanceNumber(100.12345)).toBe('100.123');
  });

  it('formats decimal number with exactly three decimal places', () => {
    expect(formatBalanceNumber(100.123)).toBe('100.123');
  });

  it('formats decimal number with one decimal place', () => {
    expect(formatBalanceNumber(100.5)).toBe('100.5');
  });

  it('adds thousand separators for large numbers', () => {
    expect(formatBalanceNumber(1234567.89)).toBe('1,234,567.89');
  });

  it('handles zero', () => {
    expect(formatBalanceNumber(0)).toBe('0.0');
  });

  it('handles negative numbers', () => {
    expect(formatBalanceNumber(-123.456)).toBe('-123.456');
  });
});

describe('[UTILS]: handleAmountConstraints', () => {
  it('handles amount within valid range', () => {
    const setAmount = vi.fn();
    const setCanContinue = vi.fn();

    handleAmountConstraints({
      value: 500,
      MAX_AMOUNT: 1000,
      MAX_POOL: 2000,
      setAmount,
      setCanContinue
    });

    expect(setAmount).toHaveBeenCalledWith(500);
    expect(setCanContinue).toHaveBeenCalledWith(true);
  });

  it('truncates amount that exceeds MAX_POOL', () => {
    const setAmount = vi.fn();
    const setCanContinue = vi.fn();

    handleAmountConstraints({
      value: 3000,
      MAX_AMOUNT: 1000,
      MAX_POOL: 2000,
      setAmount,
      setCanContinue
    });

    expect(setAmount).toHaveBeenCalledWith(2000);
    expect(setCanContinue).toHaveBeenCalledWith(false);
  });

  it('sets zero for negative values', () => {
    const setAmount = vi.fn();
    const setCanContinue = vi.fn();

    handleAmountConstraints({
      value: -100,
      MAX_AMOUNT: 1000,
      MAX_POOL: 2000,
      setAmount,
      setCanContinue
    });

    expect(setAmount).toHaveBeenCalledWith(0);
    expect(setCanContinue).toHaveBeenCalledWith(false);
  });

  it('sets canContinue to false when amount exceeds MAX_AMOUNT', () => {
    const setAmount = vi.fn();
    const setCanContinue = vi.fn();

    handleAmountConstraints({
      value: 1500,
      MAX_AMOUNT: 1000,
      MAX_POOL: 2000,
      setAmount,
      setCanContinue
    });

    expect(setAmount).toHaveBeenCalledWith(1500);
    expect(setCanContinue).toHaveBeenCalledWith(false);
  });

  it('sets canContinue to false when amount is zero', () => {
    const setAmount = vi.fn();
    const setCanContinue = vi.fn();

    handleAmountConstraints({
      value: 0,
      MAX_AMOUNT: 1000,
      MAX_POOL: 2000,
      setAmount,
      setCanContinue
    });

    expect(setAmount).toHaveBeenCalledWith(0);
    expect(setCanContinue).toHaveBeenCalledWith(false);
  });
});
