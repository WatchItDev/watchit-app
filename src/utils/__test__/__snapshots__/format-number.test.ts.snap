// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[UTILS]: fNumber > matches snapshot for different inputs 1`] = `
[
  "1,000",
  "2,500",
  "-350",
  "0",
  "0",
]
`;
